lib/app.d.ts

```diff
@@ -0,0 +1,313 @@
+        readonly providerType: string;
+
+        /**
+         * A simple object which can be passed to the server as the body of a request to authenticate.
+         */
+        readonly payload: PayloadType;
+
+        /**
+         * Creates credentials that logs in using the [Anonymous Provider](https://docs.mongodb.com/stitch/authentication/anonymous/).
+         *
+         * @returns The credentials instance, which can be passed to `app.logIn`.
+         */
+        static anonymous(): Credentials<Credentials.AnonymousPayload>;
+
+        /**
+         * Creates credentials that logs in using the [API Key Provider](https://docs.mongodb.com/stitch/authentication/api-key/).
+         *
+         * @param key The secret content of the API key.
+         * @returns The credentials instance, which can be passed to `app.logIn`.
+         */
+        static apiKey(key: string): Credentials<Credentials.ApiKeyPayload>;
+
+        /**
+         * Creates credentials that logs in using the [Email/Password Provider](https://docs.mongodb.com/stitch/authentication/userpass/).
+         * Note: This was formerly known as the "Username/Password" provider.
+         *
+         * @param email The end-users email address.
+         * @param password The end-users password.
+         * @returns The credentials instance, which can be passed to `app.logIn`.
+         */
+        static emailPassword(
+            email: string,
+            password: string,
+        ): Credentials<Credentials.EmailPasswordPayload>;
+
+        // TODO: Add providerCapabilities?
+    }
+
+    /**
+     * The constructor of MongoDB Realm App.
+     */
+    type AppConstructor = new <
+        FunctionsFactoryType extends object = DefaultFunctionsFactory
+    >(
+        idOrConfiguration: string | AppConfiguration,
+    ) => App<FunctionsFactoryType>;
+
+    /**
+     * A MongoDB Realm App.
+     */
+    class App<FunctionsFactoryType extends object = DefaultFunctionsFactory> {
+        /**
+         *
+         */
+        static readonly Credentials: typeof Credentials;
+
+        /**
+         * The id of this Realm app.
+         */
+        readonly id: string;
+
+        /**
+         * Use this to call functions defined on the MongoDB Realm server.
+         */
+        readonly functions: FunctionsFactoryType & BaseFunctionsFactory;
+
+        /**
+         * The last user to log in or being switched to.
+         */
+        readonly currentUser: Realm.User | null;
+
+        /**
+         * All authenticated users.
+         */
+        readonly allUsers: Readonly<Realm.User[]>;
+
+        /**
+         * Log in a user using a specific credential
+         *
+         * @param credentials the credentials to use when logging in
+         */
+        logIn(credentials: Credentials): Promise<Realm.User>;
+
+        /**
+         * Log out the currently authenticated user and clear any persisted authentication information.
+         */
+        logOut(): Promise<void>;
```

Comment:
Yeah - but I'll postpone that for [phase 2](https://github.com/realm/realm-js/issues/2826), since this has already shipped before I learned about that change to the API.